File: original/predicates.c
#include "copyright.h"

/* Predicates for testing various conditions */

#include <ctype.h>

#include "db.h"
#include "interface.h"
#include "config.h"
#include "externs.h"

int can_link_to(dbref who, dbref where)
{
    return(where >= 0
	   && where < db_top
	   && Typeof(where) == TYPE_ROOM
	   && (controls(who, where) || (db[where].flags & LINK_OK)));
}

int could_doit(dbref player, dbref thing)
{
    dbref key;
    int status;

    if(Typeof(thing) != TYPE_ROOM && db[thing].location == NOTHING) return 0;
    if((key = db[thing].key) == NOTHING) return 1;
    status = (player == key || member(key, db[player].contents));
    return((db[thing].flags & ANTILOCK) ? !status : status);
}

int can_doit(dbref player, dbref thing, const char *default_fail_msg)
{
    dbref loc;
    char buf[BUFFER_LEN];

    if((loc = getloc(player)) == NOTHING) return 0;

    if(!could_doit(player, thing)) {
	/* can't do it */
	if(db[thing].fail_message) {
	    notify(player, db[thing].fail_message);
	} else if(default_fail_msg) {
	    notify(player, default_fail_msg);
	}
	
	if(db[thing].ofail) {
	    sprintf(buf, "%s %s", db[player].name, db[thing].ofail);
	    notify_except(db[loc].contents, player, buf);
	}

	return 0;
    } else {
	/* can do it */
	if(db[thing].succ_message) {
	    notify(player, db[thing].succ_message);
	}

	if(db[thing].osuccess) {
	    sprintf(buf, "%s %s", db[player].name, db[thing].osuccess);
	    notify_except(db[loc].contents, player, buf);
	}

	return 1;
    }
}

int can_see(dbref player, dbref thing, int can_see_loc)
{
    if(player == thing || Typeof(thing) == TYPE_EXIT) {
	return 0;
    } else if(can_see_loc) {
	return(!Dark(thing) || controls(player, thing));
    } else {
	/* can't see loc */
	return(controls(player, thing));
    }
}

int controls(dbref who, dbref what)
{
    /* Wizard controls everything */
    /* owners control their stuff */
    return(what >= 0
	   && what < db_top
	   && (Wizard(who)
	       || who == db[what].owner));
}

int can_link(dbref who, dbref what)
{
    return((Typeof(what) == TYPE_EXIT && db[what].location == NOTHING)
	   || controls(who, what));
}

int payfor(dbref who, int cost)
{
    if(Wizard(who)) {
	return 1;
    } else if(db[who].pennies >= cost) {
	db[who].pennies -= cost;
	return 1;
    } else {
	return 0;
    }
}

int ok_name(const char *name)
{
    return (name
	    && *name
	    && *name != LOOKUP_TOKEN
	    && *name != NUMBER_TOKEN
	    && string_compare(name, "me")
	    && string_compare(name, "home")
	    && string_compare(name, "here"));
}

int ok_player_name(const char *name)
{
    const char *scan;

    if(!ok_name(name)) return 0;

    for(scan = name; *scan; scan++) {
	if(!isgraph(*scan)) {
	    return 0;
	}
    }

    /* lookup name to avoid conflicts */
    return (lookup_player(name) == NOTHING);
}

File: ruby/lib/predicates.rb
require_relative '../test/include'
require_relative '../test/defines.rb'
require_relative './helpers.rb'

module TinyMud
  class Predicates
    include Helpers

    def initialize(db)
      @db = db
    end

    def can_link_to(who, where)
        where >= 0 &&
        where < @db.length &&
        typeof(where) == TYPE_ROOM &&
        (controls(who, where) || is_link_ok(where))
    end

    def could_doit(player, thing)
      return false if (typeof(thing) != TYPE_ROOM && @db[thing].location == NOTHING)
      return true if ((key = @db[thing].key) == NOTHING)
      status = (player == key || Utils.new(@db).member(key, @db[player].contents))
      return is_antilock(thing) ? !status : status
    end
  
    def can_doit(player, thing, default_fail_msg)
      loc = getloc(player)

      return false if (loc == NOTHING)

      if (!could_doit(player, thing))
        # can't do it
        if (@db[thing].fail)
          Interface.do_notify(player, @db[thing].fail)
        elsif (default_fail_msg)
          Interface.do_notify(player, default_fail_msg)
        end
  
        if (@db[thing].ofail)
          Speech.new(@db).notify_except(@db[loc].contents, player, "#{@db[player].name} #{@db[thing].ofail}".to_s)
        end
        false
      else
        # can do it
        if (@db[thing].succ)
          Interface.do_notify(player, @db[thing].succ)
        end
    
        if (@db[thing].osucc)
          Speech.new(@db).notify_except(@db[loc].contents, player, "#{@db[player].name} #{@db[thing].osucc}")
        end
        true
      end
    end
  
    def can_see(player, thing, can_see_loc)
      if (player == thing || typeof(thing) == TYPE_EXIT)
        return false
      elsif can_see_loc
        return !is_dark(thing) || controls(player, thing)
      else
        # can't see loc
        controls(player, thing)
      end
    end
    
    def controls(who, what)
      # Wizard controls everything
      # owners control their stuff
      what >= 0 &&
      what < @db.length &&
      (is_wizard(who) || who == @db[what].owner)
    end
    
    def can_link(who, what)
      (typeof(what) == TYPE_EXIT && @db[what].location == NOTHING) || controls(who, what)
    end
    
    def payfor(who, cost)
      if (is_wizard(who))
        return true
      elsif (@db[who].pennies >= cost)
        @db[who].pennies -= cost
        return true
      else
        return false
      end
    end

    def ok_name(name)
      !name.nil? &&
      !name.empty? &&
      (name[0] != 0.chr) &&
      (name[0] != LOOKUP_TOKEN) &&
      (name[0] != NUMBER_TOKEN) &&
      (name != "me") &&
      (name != "home") &&
      (name != "here")
    end

    def ok_player_name(name)
      return false if name.nil?
      return false unless ok_name(name)
      return false if name.match(/[^[:graph:]]/)
      return Player.new(@db).lookup_player(name) == NOTHING
    end
  end
end
